python/src/trezorlib/cardano.py

```diff
@@ -390,44 +458,71 @@ def get_public_key(client, address_n: List[int]) -> messages.CardanoPublicKey:
-    withdrawals: List[messages.CardanoTxWithdrawalType] = (),
+    certificates: List[messages.CardanoTxCertificate] = (),
+    withdrawals: List[messages.CardanoTxWithdrawal] = (),
     protocol_magic: int = PROTOCOL_MAGICS["mainnet"],
     network_id: int = NETWORK_IDS["mainnet"],
-    auxiliary_data: messages.CardanoTxAuxiliaryDataType = None,
-) -> messages.CardanoSignedTx:
+    auxiliary_data: messages.CardanoTxAuxiliaryData = None,
+) -> SignTxResponse:
+    witness_paths = _get_witness_paths(inputs, certificates, withdrawals)
+
     response = client.call(
         messages.CardanoSignTx(
-            inputs=inputs,
-            outputs=outputs,
+            signing_mode=_get_signing_mode(certificates),
+            inputs_count=len(inputs),
+            outputs_count=len(outputs),
             fee=fee,
             ttl=ttl,
             validity_interval_start=validity_interval_start,
-            certificates=certificates,
-            withdrawals=withdrawals,
+            certificates_count=len(certificates),
+            withdrawals_count=len(withdrawals),
             protocol_magic=protocol_magic,
             network_id=network_id,
-            auxiliary_data=auxiliary_data,
+            has_auxiliary_data=auxiliary_data is not None,
+            witnesses_count=len(witness_paths),
         )
     )
+    if not isinstance(response, messages.CardanoTxItemAck):
+        raise exceptions.TrezorException("Unexpected response")
+
+    for tx_item in chain(inputs, _get_output_items(outputs), certificates, withdrawals):
+        response = client.call(tx_item)
+        if not isinstance(response, messages.CardanoTxItemAck):
+            raise exceptions.TrezorException("Unexpected response")
 
-    result = bytearray()
-    while isinstance(response, messages.CardanoSignedTxChunk):
-        result.extend(response.signed_tx_chunk)
-        response = client.call(messages.CardanoSignedTxChunkAck())
+    sign_tx_response = {}
 
-    if not isinstance(response, messages.CardanoSignedTx):
+    if auxiliary_data is not None:
+        response = client.call(auxiliary_data)
+        if auxiliary_data.catalyst_registration_parameters:
+            if not isinstance(response, messages.CardanoTxAuxiliaryDataSupplement):
+                raise exceptions.TrezorException("Unexpected response")
+            sign_tx_response["auxiliary_data_supplement"] = response.__dict__
+        else:
+            if not isinstance(response, messages.CardanoTxItemAck):
+                raise exceptions.TrezorException("Unexpected response")
+
+    sign_tx_response["witnesses"] = []
+    for path in witness_paths:
+        response = client.call(messages.CardanoTxWitnessRequest(path=path))
+        if not isinstance(response, messages.CardanoTxWitnessResponse):
+            raise exceptions.TrezorException("Unexpected response")
+        sign_tx_response["witnesses"].append((path, response.witness_signature))
```

Comment:
I'd find it more intuitive/more "user friendly" to pass an object with the keys "path" and "signature" rather than a tuple, to keep it explicit what we are returning. A "witness" in Cardano sense is the full structure with the public key as well and looks pretty similar to that tuple, so I'd rather not mislead the consumer of this output into thinking the full witness is being returned here